src/archive/sevenz.rs

```diff
@@ -150,7 +160,59 @@ where
         }
 
         Ok(true)
-    })?;
+    };
+
+    match password {
+        Some(password) => sevenz_rust::decompress_with_extract_fn_and_password(
+            reader,
+            output_path,
+            sevenz_rust::Password::from(password.to_str().map_err(|_| {
+                Error::InvalidPassword("7z requires that all passwords are valid UTF-8")
+            })?),
+            entry_extract_fn,
+        )?,
+        None => sevenz_rust::decompress_with_extract_fn(reader, output_path, entry_extract_fn)?,
+    }
 
     Ok(count)
 }
+
+/// List contents of `archive_path`, returning a vector of archive entries
+pub fn list_archive(
+    archive_path: &Path,
+    password: Option<&[u8]>,
+) -> impl Iterator<Item = crate::Result<FileInArchive>> {
```

Comment:
This should return `Result<impl Iterator<Item = Result<FileInArchive>>>` so you don't have to return a `vec![Err(Error::InvalidPassword(...))]`, which is a little hacky.

Reply:
> This should return `Result<impl Iterator<Item = Result<FileInArchive>>>` so you don't have to return a `vec![Err(Error::InvalidPassword(...))]`, which is a little hacky.

yes, but if in this way, we need to change all the funcs return type, is this OK ?

Reply:
changed for `sevenz` and `rar`,  do I need to change the return type for other types?

Reply:
> changed for `sevenz` and `rar`, do I need to change the return type for other types?
> 
> I mean, to keep they are the same

also done for `tar` and `zip`

Reply:
That wasn't needed but it's OK, in a scenario like this, if a function doesn't fail, it's better to keep it returning the type directly, and make only the fallible functions return `Result`.

There are other cases where you might want to keep a generic interface fallible, but it doesn't apply here since we are not a library, etc.